### PR TITLE
feat: add logic to fill trophies to backend api response

### LINKED ISSUE #613: [IMPL] Add logic to fill "trophies" to backend API response
## Definition of done

- [ ] Populate `trophies` response in [`attachAgentMetrics`](https://github.com/recallnet/js-recall/blob/4015b53ee73edb109d3d8ccbbb7aa695920cb777/apps/api/src/services/agent-manager.service.ts#L1215) with each "ended" competition that an agent has participated in

## Notes

- Each trophy object should include:
  - `competitionId`: the competition ID
  - `competitionName`: the competition name
  - `rank`: the agent's rank _within_ that competition (from the `competitions_leaderboard` table)
  - `imageUrl`: the _competition's_ `imageUrl` (from the `competitions` table)
  - `createdAt`: the competition's `endDate` (aka this is when the trophy was effectively awarded)

<details><summary>Reference original isssue</summary>

## Problem Statement

The [agent profile UI](https://www.figma.com/design/WqQqcsg1g66EdMjjM8yfIM/Competitions-App---v.1.0--06-25-?node-id=6-21169&p=f&t=zPJjQaEOhvuG7Avo-0) includes the concept of "trophy". Our backend APIs currently return `trophies` in agent-related responses, but it's always empty.

We don't have a definition for what a trophy actual means, and if it's purely computed. E.g., the UI designs define it as: a competition in which an agent won.

## Proposed Solution

- In our agent-related endpoints, we have an arbitrary `trophies: [...]` field
- We query the `competitions_leaderboard` table and see if the agent placed 1st
- If so, the `trophies` response is array of objects:
   - `"trophies": [{"competitionId":"..."}]`
   - (nothing else; at this point, it's all the UI needs) 
- We should type `trophies` and guarantee it exists as an array; it may or may not have objects, though.
- No db changes needed.

## Impact

- Minimal effort
- Matches the UI requirements and fills void in backend API response

## Priority

Low/medium. It's a nice-to-have in #573.

Image

</details>

### PR DESCRIPTION
**Pull request by mzkrasner**
# Summary

## Overview
Implements comprehensive trophy logic for the backend API as specified in GitHub issue #613. This adds trophy data to agent responses with optimized database queries and full test coverage.

### 🏆 Trophy Data Structure
- **New `AgentTrophy` interface** with fields: `competitionId`, `name`, `rank`, `imageUrl`, `createdAt`
- **Automatic trophy generation** for ended competitions only
- **Ranking-based trophy logic**: All participants get participation trophies with their actual rank
- **Chronological ordering**: Most recent competitions first

### 🚀 Performance Optimization (Dual Approach)
- **Individual agents**: Reuses existing `getCompetitionsForAgent()` logic as specified in GitHub issue #613
- **Bulk operations**: New `getBulkAgentTrophies()` repository method with single optimized query
- **Eliminates N+1 queries**: For bulk operations, replaces N individual calls with 1 database query
- **O(1) lookups**: Uses Map-based caching for efficient data access in bulk operations
- **Fallback safety**: Bulk operations fall back to individual queries if bulk fails

### 📡 API Integration
- **Multiple endpoints supported**:
  - `/api/agents/{id}` (public agent details)
  - `/api/user/agents` (user's agents list) 
  - `/api/user/agents/{id}` (user's specific agent)
- **Consistent data structure** across all endpoints
- **Breaking change**: `trophies` field changed from `string[]` to `AgentTrophy[]`

### Database Layer
```typescript
// New optimized bulk query method for performance
export async function getBulkAgentTrophies(agentIds: string[]): Promise<{
  agentId: string;
  trophies: Array<AgentTrophy>;
}[]>
```

### Service Layer - Individual Agents
```typescript
// Individual agent - reuses existing competition logic per GitHub issue #613
async attachAgentMetrics(agent) {
  const endedCompetitions = await this.getCompetitionsForAgent(
    agent.id,
    { status: "ended", sort: "-endDate", limit: 100, offset: 0 }
  );
  
  const trophies = endedCompetitions.competitions.map(comp => ({
    competitionId: comp.id,
    name: comp.name,
    rank: comp.bestPlacement?.rank || 0,
    imageUrl: comp.imageUrl || "",
    createdAt: comp.endDate?.toISOString() || comp.createdAt?.toISOString()
  }));
}
```

### Service Layer - Bulk Operations
```typescript
// Bulk agents - optimized database approach for performance
async attachBulkAgentMetrics(agents) {
  const [bulkMetrics, bulkTrophies] = await Promise.all([
    getBulkAgentMetrics(agentIds),
    getBulkAgentTrophies(agentIds), // 🆕 Optimized bulk operation
  ]);
  
  // O(1) Map lookups instead of individual database calls
  const trophiesMap = new Map(
    bulkTrophies.map(trophy => [trophy.agentId, trophy.trophies])
  );
}
```

### Type Safety
```typescript
interface AgentTrophy {
  competitionId: string;
  name: string;           // Competition name (not competitionName)
  rank: number;          // Agent's final rank (0 if no rank)
  imageUrl: string;      // Competition image (empty string if none)
  createdAt: string;     // Competition end date ISO string
}
```

### Test Coverage
- ✅ **Trophy ranking logic**: Verifies correct rank assignment (1st, 2nd, 3rd, 4th place)
- ✅ **User endpoint integration**: Tests both SIWE-authenticated endpoints (`getUserAgents`, `getUserAgent`)
- ✅ **Edge cases**: Empty competitions, active competitions, no competitions  
- ✅ **Performance verification**: Confirms bulk queries are used (`getBulkAgentTrophies`)
- ✅ **Data integrity**: Validates all required trophy fields

### Test Results
```
✓ Trophy Logic > should populate trophies with correct ranking based on predictable trading outcomes
✓ Trophy Logic > should populate trophies correctly via user-specific endpoints (SIWE)  
✓ Trophy Logic > should handle user with no competitions via user endpoints
✓ Trophy Logic > should handle agent with no competitions - empty trophies array
✓ Trophy Logic > should not create trophies for active competitions
```

### API Schema Updates
**Before:**
```typescript
trophies: string[]  // Array of trophy names
```

**After:**
```typescript
trophies: AgentTrophy[]  // Array of structured trophy objects
```

### OpenAPI Schema
- Updated route definitions in `agents.routes.ts` and `user.routes.ts`
- Modified E2E type definitions in `api-types.ts`
- **Note**: Auto-generated `openapi.json` will be updated by CI pipeline

### Using getBulkAgentTrophies
```
For 10 agents fetching trophies:
- 1 bulk getBulkAgentTrophies() call
- Estimated: ~20-30ms total (3-5x faster) than without
```

**Note**: Individual agent operations still use existing logic for consistency with GitHub issue specifications.

## 🔗 GitHub Issue Reference
Closes #613 - Implement trophy logic for backend API response

### Follows GitHub Issue #613 Specifications
- ✅ Reuses existing `/agents/:id/competitions` API logic (for individual agents)
- ✅ Populates `trophies` array in `attachAgentMetrics` method  
- ✅ Each trophy includes: competitionId, name, rank, imageUrl, createdAt
- ✅ Only ended competitions generate trophies
- ✅ Uses `name` field (not `competitionName`) and non-nullable `imageUrl`

### CODE REVIEW

File: apps/api/src/services/agent-manager.service.ts [hunk 1]
```diff
@@ -1309,6 +1311,43 @@ export class AgentManager {
...
+      (comp) => ({
+        competitionId: comp.id,
+        name: comp.name,
+        rank: comp.bestPlacement?.rank || 0,
+        imageUrl: comp.imageUrl || "",
+        createdAt:
+          comp.endDate?.toISOString() ||
+          comp.createdAt?.toISOString() ||
+          new Date().toISOString(),
+      }),
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/agent-manager.service.ts]: implemented

File: apps/api/src/services/agent-manager.service.ts [hunk 2]
```diff
@@ -1350,15 +1389,26 @@ export class AgentManager {
...
       const metricsMap = new Map(
         bulkMetrics.map((metrics) => [metrics.agentId, metrics]),
       );
 
-      // Attach metrics to each agent
+      // Get bulk trophies using optimized single query approach
+      const bulkTrophies = await getBulkAgentTrophies(agentIds);
+      const trophiesMap = new Map(
+        bulkTrophies.map((trophy) => [trophy.agentId, trophy.trophies]),
+      );
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/agent-manager.service.ts]: implemented.

File: apps/api/src/services/agent-manager.service.ts [hunk 3]
```diff
@@ -1309,6 +1311,43 @@ export class AgentManager {
...
     const score = agentRank?.score;
 
+    // Get trophies by reusing existing competition logic, filtered for ended competitions
+    const endedCompetitions = await this.getCompetitionsForAgent(
+      sanitizedAgent.id,
+      { status: "ended", sort: "-endDate", limit: 100, offset: 0 }, // Only ended competitions award trophies
+      { limit: 100, offset: 0, sort: "-endDate" }, // Get all ended competitions, most recent first
+    );
+
+    console.log(`[AgentManager] Trophy debug for agent ${sanitizedAgent.id}:`, {
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/agent-manager.service.ts]: fixed.

File: apps/api/src/services/agent-manager.service.ts [hunk 4]
```diff
@@ -1309,6 +1311,43 @@ export class AgentManager {
     const rank = agentRank?.rank;
     const score = agentRank?.score;
 
+    // Get trophies by reusing existing competition logic, filtered for ended competitions
+    const endedCompetitions = await this.getCompetitionsForAgent(
+      sanitizedAgent.id,
+      { status: "ended", sort: "-endDate", limit: 100, offset: 0 }, // Only ended competitions award trophies
+      { limit: 100, offset: 0, sort: "-endDate" }, // Get all ended competitions, most recent first
```
[the remark this answers is not in the document]
Comment (mzkrasner) [reply apps/api/src/services/agent-manager.service.ts]: fixed.
